File: frontend/src/services/ChatService.js
import BaseService from './BaseService.js';
import Messaggio from '../models/Messaggio.js';
import Utente from '../models/Utente.js';

/**
 * Service per la chat: storico dei messaggi, contatti disponibili e gestione dei
 * messaggi non letti, con mappatura sui modelli {@link Messaggio} e {@link Utente}.
 */
class ChatService extends BaseService {

  async getStorico(altroUtenteId) {
    return Messaggio.fromList(await this.get(`/chat/storico/${altroUtenteId}`));
  }

  async getContatti() {
    return Utente.fromList(await this.get('/chat/contatti'));
  }

  // Mappa { mittenteId: numeroNonLetti }
  getNonLetti() {
    return this.get('/chat/non-letti');
  }

  segnaLetti(altroUtenteId) {
    return this.put(`/chat/${altroUtenteId}/letti`);
  }
}

export default new ChatService();

File: backend/src/main/java/com/tesi/gestionalec/service/interfaces/ChatService.java
package com.tesi.gestionalec.service.interfaces;

import com.tesi.gestionalec.dto.request.MessaggioChatRequest;
import com.tesi.gestionalec.dto.response.MessaggioChatResponse;
import com.tesi.gestionalec.dto.response.UtenteResponse;
import com.tesi.gestionalec.model.Utente;

import java.util.List;
import java.util.Map;

/**
 * Contratto per la messaggistica interna: invio e storico dei messaggi,
 * individuazione dei contatti ammessi per ruolo e gestione dei messaggi non letti.
 */
public interface ChatService {
    MessaggioChatResponse salvaEInvia(MessaggioChatRequest request, Utente mittente);
    List<MessaggioChatResponse> storico(Long utenteAId, Long utenteBId);
    /** Restituisce i contatti con cui l'utente può chattare in base al suo ruolo. */
    List<UtenteResponse> trovaContatti(Utente utente);
    /** Messaggi non letti dell'utente raggruppati per id del mittente. */
    Map<Long, Long> nonLettiPerMittente(Long utenteId);
    /** Segna come letti i messaggi ricevuti da un certo mittente. */
    void segnaLetti(Long utenteId, Long altroUtenteId);
}